### ISSUE TITLE
Common test vectors for client implementations

### ISSUE THREAD

**Issue by aes** [Unleash/unleash#217]
Create examples of configuration, queries and results, to ensure that clients behave identically.

Ensure identical behavior in case of
- different existing clients
- new clients
- string and encoding differences
- things we haven't thought of

**Comment by ivarconr** [comment]
Yes, this would be very nice to have in place. Personally I have been thinking about setting up a test-platform which could be used to verify correctness of client implementations.

Do you have concrete experiences or suggestion on how this could be implemented?

**Comment by aes** [comment]
Ok, how about this:
- First, just save a call to `/api/features`, and then
- "a0".."a10" should give [True, False, False, ...]
- "b0".."b99" should give [0xaf, 0x9d, ...]
- "c0".."c9999" should checksum to 0xdecafbad and be split [2497, 7503]

Each serves level as serves as a guide to correct implementation of the next level, and the last verifies the statistical split.

Then some examples of groupId salt, to make sure encoding craziness checks out.

**Comment by ivarconr** [comment]
Do you think it is possible to automate this? Kind of having a "unleash-verify-service" which challenges the client and the client implementation needs to provide answers back? 

Or is it better/easier to just document this in a step-by-step guide?

**Comment by aes** [comment]
Here's what I have so far:
https://github.com/aes/unleash-client-python/blob/master/tests/test_strategy.py

This is roughly the type of known test-vectors I was thinking of, and they should be relatively easy to port to wherever. (Not the random, because random is implementation-dependent, but the hash tests.)

**Comment by SimenB** [comment]
This is awesome! This isn't helping (I'm just curious), but how would this be invoked from the different clients?

**Comment by aes** [comment]
Well, you'd set up the same config, and check that you get the same results. I've only written unit tests, but it should probably be config -> context-vector -> results-vector, similar to this, but more explicitly tied to a concrete config

**Comment by ivarconr** [comment]
Nice, this looks promising! Providing test-vectors for the various strategy definitions defensively makes sense. I also believe one could do this for all parts of an unleash client (metrics, registration, etc).

**Comment by ivarconr** [comment]
After #247 it should be trivial to create common test-vectors for all standardised strategies. 

This should probably be defined as a set of test-cases all clients should run as e2e tests. Each test-case should provide a set of feature toggles, with configues strategies, unleashContext details and the expected output (true or false)

**Comment by ivarconr** [comment]
I have started on this in a separate repo: https://github.com/Unleash/client-specification

- java-impl: https://github.com/Unleash/unleash-client-java/compare/master...integration-spec
- node-impl: https://github.com/Unleash/unleash-client-node/compare/client-specification?expand=1

**Comment by ivarconr** [comment]
So this just happened, defining test cases for built-in strategies and expected results for given Unleash context.
